metagraph/benchmarks/graph/bench_boss.cpp

```diff
@@ -14,17 +15,47 @@ using mtg::graph::boss::BOSS;
 const std::string filename = "./benchmark_graph.dbg";
 
 constexpr uint64_t NUM_DISTINCT_INDEXES = 1 << 21;
+constexpr uint64_t PATH_SIZE = 1 << 7;
+constexpr uint64_t CACHE_SIZE = 1 << 10;
+
+std::shared_ptr<DBGSuccinctCachedView>
+make_cached_dbgsuccinct(std::shared_ptr<const DBGSuccinct> graph, size_t cache_size = 1024) {
+    if (graph->get_k() * mtg::kmer::KmerExtractorBOSS::bits_per_char <= 64) {
+        return std::make_shared<DBGSuccinctCachedViewImpl<mtg::kmer::KmerExtractorBOSS::Kmer64>>(graph, cache_size);
+    } else if (graph->get_k() * mtg::kmer::KmerExtractorBOSS::bits_per_char <= 128) {
+        return std::make_shared<DBGSuccinctCachedViewImpl<mtg::kmer::KmerExtractorBOSS::Kmer128>>(graph, cache_size);
+    } else {
+        return std::make_shared<DBGSuccinctCachedViewImpl<mtg::kmer::KmerExtractorBOSS::Kmer256>>(graph, cache_size);
+    }
+}
 
-
-std::unique_ptr<DBGSuccinct> load_graph(benchmark::State &state) {
-    auto graph = std::make_unique<DBGSuccinct>(2);
+template <typename Graph = DBGSuccinct, typename OutGraph = DBGSuccinct>
+std::shared_ptr<OutGraph> load_graph(benchmark::State &state) {
+    auto graph = std::make_shared<DBGSuccinct>(2);
     if (!std::getenv("GRAPH")) {
         state.SkipWithError("Set environment variable GRAPH");
     } else if (!graph->load(std::getenv("GRAPH"))) {
         state.SkipWithError((std::string("Can't load the graph from ")
                                 + std::getenv("GRAPH")).c_str());
     }
-    return graph;
+
+    if constexpr(std::is_same_v<OutGraph, DBGSuccinct>) {
+        return graph;
+    } else {
+        static_assert(std::is_same_v<OutGraph, DeBruijnGraph>);
+        if constexpr(std::is_same_v<Graph, CanonicalDBG>) {
+            return std::make_shared<CanonicalDBG>(
+                std::dynamic_pointer_cast<DeBruijnGraph>(graph)
+            );
+        } else if constexpr(std::is_same_v<Graph, DBGSuccinctCachedView>) {
+            std::shared_ptr<DeBruijnGraph> wrapped_graph
+                = make_cached_dbgsuccinct(graph, CACHE_SIZE);
+            if (graph->get_mode() == DeBruijnGraph::PRIMARY)
+                wrapped_graph = std::make_shared<CanonicalDBG>(wrapped_graph);
+
+            return wrapped_graph;
+        }
+    }
```

Comment:
Oh, this is really weird construct. Keep the old function. It was simple and interpretable.

Reply:
I split this up, so now there are two methods `load_graph` and `wrap_graph`